File: src/urich/discovery/protocol.py
"""Service Discovery protocol: resolve(service_name) -> URL(s)."""
from typing import Protocol, runtime_checkable


@runtime_checkable
class ServiceDiscovery(Protocol):
    """
    How to resolve services by name. User chooses implementation (config, Consul, etcd).
    """

    def resolve(self, service_name: str) -> list[str]:
        """Return list of URLs (e.g. one for static)."""
        ...


def static_discovery(services: dict[str, str]) -> ServiceDiscovery:
    """Minimal out-of-the-box implementation: name -> URL map."""
    return StaticDiscovery(services)


class StaticDiscovery:
    """Discovery from static config (env/map)."""

    def __init__(self, services: dict[str, str]) -> None:
        self._services = dict(services)

    def resolve(self, service_name: str) -> list[str]:
        url = self._services.get(service_name)
        return [url] if url else []

File: urich-rs/src/discovery/protocol.rs
//! Service Discovery implementations. Trait is in core::service_discovery.

use urich_core::ServiceDiscovery;

/// Discovery from static config (name -> URL map). Like Python StaticDiscovery.
#[derive(Clone, Default)]
pub struct StaticDiscovery {
    services: std::collections::HashMap<String, String>,
}

impl StaticDiscovery {
    pub fn new(services: std::collections::HashMap<String, String>) -> Self {
        Self { services }
    }

    /// Build from a slice of (name, url) pairs.
    pub fn from_slice(pairs: &[(&str, &str)]) -> Self {
        Self {
            services: pairs
                .iter()
                .map(|(k, v)| ((*k).to_string(), (*v).to_string()))
                .collect(),
        }
    }
}

impl ServiceDiscovery for StaticDiscovery {
    fn resolve(&self, service_name: &str) -> Vec<String> {
        self.services
            .get(service_name)
            .map(|u| vec![u.clone()])
            .unwrap_or_default()
    }
}
